pallets/phala/src/mining.rs

```diff
@@ -531,28 +553,102 @@ pub mod pallet {
 		}
 	}
 
+	struct Tokenomic<T> {
+		params: TokenomicParams,
+		mark: PhantomData<T>,
+	}
+
+	impl<T> Tokenomic<T>
+	where
+		T: Config,
+		BalanceOf<T>: FixedPointConvert,
+	{
+		fn new(params: TokenomicParams) -> Self {
+			Tokenomic {
+				params,
+				mark: Default::default(),
+			}
+		}
+
+		/// Gets the minimal stake with the given performance score
+		fn minimal_stake(&self, p: u32) -> BalanceOf<T> {
+			let p = FixedPoint::from_num(p);
+			let k = FixedPoint::from_bits(self.params.k);
+			let min_stake = k * p.sqrt();
+			FixedPointConvert::from_fixed(&min_stake)
+		}
+
+		/// Calcuates the initial Ve
+		fn ve(&self, s: BalanceOf<T>, p: u32, confidence_level: u8) -> FixedPoint {
+			let f1 = FixedPoint::from_num(1);
+			let score = Self::confidence_score(confidence_level);
+			let re = FixedPoint::from_bits(self.params.re);
+			let tweaked_re = (re - f1) * score + f1;
+			let s = s.to_fixed();
+			let c = self.rig_cost(p);
+			tweaked_re * (s + c)
+		}
+
+		/// Gets the max v in fixed point
+		fn v_max(&self) -> FixedPoint {
+			FixedPoint::from_bits(self.params.v_max)
+		}
+
+		/// Gets the estimated rig costs in PHA
+		fn rig_cost(&self, p: u32) -> FixedPoint {
+			let cost_k = FixedPoint::from_bits(self.params.rig_k);
+			let cost_b = FixedPoint::from_bits(self.params.rig_b);
+			let pha_rate = FixedPoint::from_bits(self.params.pha_rate);
+			let p = FixedPoint::from_num(p);
+			(cost_k * p + cost_b) / pha_rate
+		}
+
+		/// Gets the operating cost per sec
+		#[cfg(test)]
+		fn op_cost(&self, p: u32) -> FixedPoint {
+			let cost_k = FixedPoint::from_bits(self.params.cost_k);
+			let cost_b = FixedPoint::from_bits(self.params.cost_b);
+			let pha_rate = FixedPoint::from_bits(self.params.pha_rate);
+			let p = FixedPoint::from_num(p);
+			(cost_k * p + cost_b) / pha_rate
+		}
+
+		/// Converts confidence level to score
+		fn confidence_score(confidence_level: u8) -> FixedPoint {
+			const SCORES_BASE10: [u32; 5] = [10, 10, 10, 8, 7];
+			let score_base10 = if 1 <= confidence_level && confidence_level <= 5 {
+				SCORES_BASE10[confidence_level as usize - 1]
+			} else {
+				10u32
```

Comment:
Should all the other level have a best score?
By the way, since we have `fixed-macro` now, we can simply express then as `fixed!(0.8)`?

Reply:
There's only 1-5 defined. All the rests should be undefined, and just reserved for testing purpose. So we give 1 as a fallback.

fixed_macro applied.